solls/LanguageServer.cpp

```diff
@@ -0,0 +1,668 @@
+				_position.line,
+				_position.column);
+
+	return closestMatch;
+}
+
+void LanguageServer::operator()(lsp::protocol::DefinitionParams const& _params)
+{
+	lsp::protocol::DefinitionReplyParams output{};
+
+	if (auto const file = m_vfs.find(_params.textDocument.uri); file != nullptr)
+	{
+		compile(*file);
+		solAssert(m_compilerStack.get() != nullptr, "");
+
+		auto const sourceName = file->uri().substr(7); // strip "file://"
+
+		if (auto const sourceNode = findASTNode(_params.position, sourceName); sourceNode)
+		{
+			if (auto const importDirective = dynamic_cast<ImportDirective const*>(sourceNode))
+			{
+				// When cursor is on an import directive, then we want to jump to the actual file that
+				// is being imported.
+				if (auto const fpm = m_fileReader->fullPathMapping().find(importDirective->path()); fpm != m_fileReader->fullPathMapping().end())
+				{
+					output.uri = "file://" + fpm->second;
+					reply(_params.requestId, output);
+				}
+				else
+					error(_params.requestId, lsp::protocol::ErrorCode::InvalidParams, "Declaration not found.");
+			}
+			else if (auto const declaration = dynamic_cast<Declaration const*>(sourceNode))
+			{
+				// For any kind of declaration, jump to the referencing symbol of that declaration.
+				if (auto const loc = declarationPosition(declaration); loc.has_value())
+				{
+					output.range = loc.value();
+					output.uri = "file://" + declaration->location().source->name();
+					reply(_params.requestId, output);
+				}
+				else
+					error(_params.requestId, lsp::protocol::ErrorCode::InvalidParams, "Declaration not found.");
+			}
+			else if (auto const n = dynamic_cast<frontend::MemberAccess const*>(sourceNode); n)
+			{
+				// For scope members, jump to the naming symbol of the referencing declaration of this member.
+				auto const declaration = n->annotation().referencedDeclaration;
+
+				if (auto const loc = declarationPosition(declaration); loc.has_value())
+				{
+					auto const sourceName = declaration->location().source->name();
+					auto const fullSourceName = m_fileReader->fullPathMapping().at(sourceName);
+					output.range = loc.value();
+					output.uri = "file://" + fullSourceName;
+					reply(_params.requestId, output);
+				}
+				else
+					error(_params.requestId, lsp::protocol::ErrorCode::InvalidParams, "Declaration not found.");
+			}
+			else if (auto const sourceIdentifier = dynamic_cast<Identifier const*>(sourceNode); sourceIdentifier != nullptr)
+			{
+				// For identifiers, jump to the naming symbol of the definition of this identifier.
+				auto const declaration = !sourceIdentifier->annotation().candidateDeclarations.empty()
+					? sourceIdentifier->annotation().candidateDeclarations.front()
+					: sourceIdentifier->annotation().referencedDeclaration;
+
+				if (auto const loc = declarationPosition(declaration); loc.has_value())
+				{
+					output.range = loc.value();
+					output.uri = "file://" + declaration->location().source->name();
```

Comment:
There is a lot of `file://` prefixing and removing. Maybe it is worth two helpers and/or overloading?